src/solvers/sat/satcheck_ipasir.cpp

```diff
@@ -0,0 +1,200 @@
+/*******************************************************************\
+
+Module:
+
+Author: Norbert Manthey, nmanthey@amazon.com
+
+\*******************************************************************/
+
+#ifndef _MSC_VER
+#include <inttypes.h>
+#endif
+
+#include <cassert>
+#include <stack>
+
+#include <util/threeval.h>
+
+#include "satcheck_ipasir.h"
+
+#ifdef HAVE_IPASIR
+
+extern "C"
+{
+#include <ipasir.h>
+}
+
+/* solver interface:
+
+Representation:
+Variables for a formula start with 1! 0 is used as termination symbol.
+
+Functions:
+const char * ipasir_signature ();
+void * ipasir_init ();
+void ipasir_release (void * solver);
+void ipasir_add (void * solver, int lit_or_zero);
+void ipasir_assume (void * solver, int lit);
+int ipasir_solve (void * solver);
+int ipasir_val (void * solver, int lit);
+int ipasir_failed (void * solver, int lit);
+void ipasir_set_terminate (void * solver, void * state, int (*terminate)(void * state));
+
+*/
+
+tvt satcheck_ipasirt::l_get(literalt a) const
+{
+  if(a.is_true())
+    return tvt(true);
+  else if(a.is_false())
+    return tvt(false);
+
+  tvt result;
+
+  // compare to internal no_variables number
+  if(a.var_no()>=(unsigned)no_variables())
+    return tvt::unknown();
+
+  const int val=ipasir_val(solver, a.var_no());
+
+  if(val>0)
+    result=tvt(true);
+  else if(val<0)
+    result=tvt(false);
+  else
+    return tvt::unknown();
+
+  if(a.sign())
+    result=!result;
+
+  return result;
+}
+
+const std::string satcheck_ipasirt::solver_text()
+{
+  return std::string(ipasir_signature());
+}
+
+void satcheck_ipasirt::lcnf(const bvt &bv)
+{
+  forall_literals(it, bv)
+  {
+    if(it->is_true())
+      return;
+    else if(!it->is_false())
+      INVARIANT(it->var_no()<(unsigned)no_variables(),
+             "reject out of bound variables");
+  }
+
+
+  forall_literals(it, bv)
+  {
+    if(!it->is_false())
+    {
+      // add literal wit correct sign
+      ipasir_add(solver, it->sign()?-it->var_no():it->var_no());
+    }
+  }
+  ipasir_add(solver, 0); // terminate clause
+
+  clause_counter++;
+}
+
+propt::resultt satcheck_ipasirt::prop_solve()
+{
+  INVARIANT(status!=statust::ERROR, "there cannot be an error");
+
+  {
+    messaget::status() <<
+      (no_variables()-1) << " variables, " <<
+      clause_counter << " clauses" << eom;
+  }
+
+  // use the internal representation, as ipasir does not support reporting the
+  // status
+  if(status==statust::UNSAT)
+  {
+    messaget::status() <<
+      "SAT checker inconsistent: instance is UNSATISFIABLE" << eom;
+  }
+  else
+  {
+    // if assumptions contains false, we need this to be UNSAT
+    bool has_false=false;
+
+    forall_literals(it, assumptions)
+      if(it->is_false())
+        has_false=true;
```

Comment:
I think ideally this should use `std::find` as it is not really using the iterator, but I'm not going to insist on that.